### PR TITLE
feat: use Cancellation object instead of unsubscribe methods

### PR DESCRIPTION
**Pull request by ttypic**
In this PR, I refactored the subscription pattern to align with the JavaScript SDK. Initially, I thought having two separate methods for subscribing and unsubscribing would be more idiomatic for Kotlin/Java. However, it seems that modern libraries have started adopting the cancellation object approach.

Now, instead of using two methods, a cancellation object is returned upon subscription, which can be used to unsubscribe. This simplifies the API and introduces a unified cancellation interface, making it consistent across the codebase. It also lays the groundwork for adding convenient features later, like the `use` method in Kotlin for cleaner syntax and resource management.

## Summary by CodeRabbit

- **New Features**
	- Introduced a `Subscription` interface for managing subscriptions and event listeners, enhancing resource management.
	- Updated listener registration methods to return a `Subscription` object, allowing for easier cancellation of subscriptions.

- **Bug Fixes**
	- Removed redundant `unsubscribe` methods across multiple interfaces to streamline listener management.

- **Refactor**
	- Consolidated listener management into single methods, improving clarity and usability in interfaces like `ConnectionStatus`, `EmitsDiscontinuities`, and others.

### CODE REVIEW

File: chat-android/src/main/java/com/ably/chat/Cancellation.kt [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+
+/**
+ * A cancellation handle, returned by various functions (mostly subscriptions)
+ * where cancellation is required.
+ */
+fun interface Cancellation {
+    /**
+     * Handle cancellation (unsubscribe listeners, clean up)
+     */
+    fun cancel()
```
Comment (AndyTWF) [inline chat-android/src/main/java/com/ably/chat/Cancellation.kt:11]: In JS we use `unsubscribe` or `off` but happy to keep `cancel` if that's more kotlin-idiomatic.

**Review by AndyTWF — APPROVED** [review]

**Comment by sacOO7** [comment]
@ttypic I am skeptical about use of `Cancellation object`. Fact is, this comes from multithreading/synchronization concept and used for managing thread/coroutine synchronization. We have similar concept in 
1. Dotnet -> [Cancellation Token](https://medium.com/@dayanandthombare/cancellation-tokens-in-net-core-b02f10024d4f)
2. Go -> [Context with cancellation](https://golangbot.com/context-timeout-cancellation/#context-with-cancellation)

While both of those languages are pretty old and natively support passing cancellation tokens, those cancel tokens are only used to manage dotnet tasks / goroutines. In both ably-dotnet and ably-go, we return `unsubscribe` function to de-register listeners.
In my honest opinion, this will cause naming asymmetry across SDKs and eventually lead to more confusion. 
PS. It is confusing for me too : (

**Comment by ttypic** [comment]
@sacOO7 Not sure what are you suggesting? And I don't see why we have to have 2 different functional interfaces for unsubscription (one with `unsubscribe()` another with `off()`).  It's not a java or kotlin way. Single interface for subscription cancellation will help us generate extensions for coroutines easier later.  

Also I expect that chat sdk users won't use this methods at all, we'll provide extensions for coroutines and maybe some other reactive frameworks.

**Comment by sacOO7** [comment]
> @sacOO7 Not sure what are you suggesting? And I don't see why we have to have 2 different functional interfaces for unsubscription (one with `unsubscribe()` another with `off()`). It's not a java or kotlin way. Single interface for subscription cancellation will help us generate extensions for coroutines easier later.
> 
> Also I expect that chat sdk users won't use this methods at all, we'll provide extensions for coroutines and maybe some other reactive frameworks.

Okay, can we use some other keyword instead of `cancel`? If we are calling subscribe to add `listener` at top level, it will return `unsubscribe` to cancel the subscription. Use of `cancel` keyword is confusing in a way that, it used to manage/cancel coroutines on one hand and here we are using it for ably specific operations.

**Pull request merged**